src/hidimstat/stat_tools.py

```diff
@@ -390,3 +390,73 @@ def two_sided_pval_from_pval(pval, one_minus_pval=None, distrib="norm"):
     )
 
     return two_sided_pval, two_sided_pval_corr
+
+
+def step_down_max_t(stat, permutation_stats):
+    """
+    Step-down maxT multiple testing procedure
+
+    | This algorithm for computing adjusted p-values :footcite:t:`westfall1993resampling`.
+    | It assumes that the test statistics is centered around zero.
+    | This algorithm controls the family-wise error rate (FWER) by:
+    | 1. Ordering test statistics by absolute value
+    | 2. Computing successively larger null distributions
+    | 3. Adjusting p-values using step-down procedure
+
+    Parameters
+    ----------
+    stat : ndarray, shape (n_features,)
+        Statistic computed on the original (unpermutted) problem.
+
+    permutation_stats : ndarray, shape (n_permutations, n_features)
+        Statistics computed on permutted problems.
+
+    Returns
+    -------
+    two_sided_pval_corr : ndarray, shape (n_features,)
+        Two-sided p-values corrected for multiple testing.
+
+    References
+    ----------
+    .. footbibliography::
+    """
+
+    n_permutations, n_features = np.shape(permutation_stats)
+
+    # Step 1: Order features by absolute value of test statistics
+    # Keep track of original positions
+    index_ordered = np.argsort(np.abs(stat))
+    stat_ranked = np.empty(n_features, dtype=int)
+    stat_ranked[index_ordered] = np.arange(n_features)
+    # Sorted absolute statistics
+    stat_sorted = np.copy(np.abs(stat)[index_ordered])
+    # Order permutation stats similarly
+    permutation_stats_ordered = np.copy(np.abs(permutation_stats)[:, index_ordered])
+
+    # Step 2: Update null distribution
+    # For each column i, take the maximum between current and previous column
+    # This creates successively larger null distributions
+    for i in range(1, n_features):
+        permutation_stats_ordered[:, i] = np.maximum(
+            permutation_stats_ordered[:, i - 1], permutation_stats_ordered[:, i]
+        )
+
+    # Step 3: Compute raw adjusted p-values
+    # Count how many permutation statistics are >= than observed statistics
+    two_sided_pval_corr = (
+        np.sum(np.less_equal(stat_sorted, permutation_stats_ordered), axis=0)
+        / n_permutations
+    )
+
+    # Step 4: Enforce monotonicity
+    # Ensure that p-values are monotonically decreasing
+    # by taking maximum of current and next p-value
+    for i in range(n_features - 1)[::-1]:
+        two_sided_pval_corr[i] = np.maximum(
+            two_sided_pval_corr[i], two_sided_pval_corr[i + 1]
+        )
+
```

Comment:
It seems to me that this algorithm is uselessly complex. Why not simply create the sorted max distribution of permutation stats, which creates a "max-null" distribution and then compare the actual statistic to this max-null distribution. This would be simpler and faster ?

Reply:
I am new to the type of algorithm, but if you think it's better, I can provide this new function. 
However, we need to be careful because, in the original book, they explain that this method is not intuitive.

Reply:
@bthirion